lib/rebaser-server/src/change_set_processor_task.rs

```diff
@@ -0,0 +1,420 @@
+                    .on_response(DefaultOnResponse::new().level(Level::TRACE)),
+            )
+            .layer(PostProcessLayer::new().on_success(DeleteMessageOnSuccess::new(stream)))
+            .service(handlers::default.with_state(state));
+
+        let inner =
+            naxum::serve_with_incoming_limit(inactive_aware_incoming, app.into_make_service(), 1)
+                .with_graceful_shutdown(graceful_shutdown_signal(
+                    task_token,
+                    quiescence_token.clone(),
+                ));
+
+        let inner_fut = inner.into_future();
+
+        Self {
+            _metadata: metadata,
+            workspace_id,
+            change_set_id,
+            inner: Box::new(inner_fut),
+            quiescence_token,
+        }
+    }
+
+    pub(crate) async fn try_run(self) -> Result<Shutdown> {
+        self.inner.await.map_err(Error::Naxum)?;
+
+        if self.quiescence_token.is_cancelled() {
+            debug!(
+                task = Self::NAME,
+                si.workspace.id = %self.workspace_id,
+                si.change_set.id = %self.change_set_id,
+                "shutdown due to quiescent period",
+            );
+            Ok(Shutdown::Quiesced)
+        } else {
+            debug!(
+                task = Self::NAME,
+                si.workspace.id = %self.workspace_id,
+                si.change_set.id = %self.change_set_id,
+                "shutdown complete",
+            );
+            Ok(Shutdown::Graceful)
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum Shutdown {
+    Graceful,
+    Quiesced,
+}
+
+struct QuiescedCaptured {
+    instance_id: String,
+    workspace_id: WorkspacePk,
+    change_set_id: ChangeSetId,
+    quiescence_token: CancellationToken,
+}
+
+#[derive(Clone, Debug)]
+struct DeleteMessageOnSuccess {
+    stream: jetstream::stream::Stream,
+}
+
+impl DeleteMessageOnSuccess {
+    fn new(stream: jetstream::stream::Stream) -> Self {
+        Self { stream }
+    }
+}
+
+impl post_process::OnSuccess for DeleteMessageOnSuccess {
+    fn call(
+        &mut self,
+        head: Arc<naxum::Head>,
+        info: Arc<post_process::Info>,
+    ) -> BoxFuture<'static, ()> {
+        let stream = self.stream.clone();
+
+        Box::pin(async move {
+            trace!("deleting message on success");
+            if let Err(err) = stream.delete_message(info.stream_sequence).await {
+                warn!(
+                    si.error.message = ?err,
+                    subject = head.subject.as_str(),
+                    "failed to delete the message",
+                );
+            }
+        })
+    }
+}
+
+// Await either a graceful shutdown signal from the task or an inactive request stream trigger.
+async fn graceful_shutdown_signal(
+    task_token: CancellationToken,
+    quiescence_token: CancellationToken,
+) {
+    tokio::select! {
+        _ = task_token.cancelled() => {}
+        _ = quiescence_token.cancelled() => {}
+    }
+}
```

Comment:
Happy that this is its own function too. Good idea

Reply:
Me too!